POLIO-1540 Create new tab sub activities for all campaigns polio and non polio

mathvdh opened the pull request:
Introduce SubActivity Model in Polio Plugin and the related APIs

Related JIRA tickets : POLIO-1540

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] My migrations file are included
- [x] Are there enough tests
- [x] Documentation has been included (for new feature)

## Doc

In the `docs/pages/dev/campaigns/subactivities.md` file included in this PR.

## Changes

- Added a new model `SubActivity` which represents a sub-activity within a round of a campaign. It has fields for name, age unit, minimum age, maximum age, start date, and end date.
- Added `SubActivityScope` model to represent the scope of a SubActivity with respect to vaccines.
- The `SubActivity` model is linked to the `Round` model via a foreign key, allowing each round to have multiple sub-activities.
- The `SubActivityScope` model is linked to the `SubActivity` model via a foreign key, allowing each sub-activity to have multiple scopes.

## How to test

As this only includes the APIs and no frontend code, you can use /swagger to test and/or use django admin.

mathvdh committed:
Initial Commit

mathvdh committed:
POLIO-1540 Add tests

mathvdh committed:
POLIO-1540 Cleanup

mathvdh committed:
POLIO-1540 Better default

mathvdh committed:
POLIO-1540 add model fields and cleanup migrations

mathvdh committed:
Adding docs

mathvdh committed:
POLIO-1540 Fix black fomatting

quang-le committed:
Merge remote-tracking branch 'origin/POLIO-1540-create-new-tab-sub-activities-for-all-campaigns-polio-and-non-polio' into POLIO-1556_sub_activities_front_end

quang-le committed:
POLIO-1556: move Evaluation files out of rounds folder

quang-le commented on plugins/polio/api/campaigns/subactivities.py:
```diff
@@ -0,0 +1,121 @@
+class SubActivityCreateUpdateSerializer(serializers.ModelSerializer):
+    round_number = serializers.IntegerField(write_only=True, required=False)
+    campaign = serializers.CharField(write_only=True, required=False)
+    scopes = SubActivityScopeSerializer(many=True, required=False)
```
shouldn't the scopes be mandatory?

madewulf replied:
I think not, you could have first just the encoding of dates and reason of activity and the scope coming later

quang-le reviewed (commented):
I tried the POST and PUT in the front-end, it works fine, but shouldn't we check the values against those of the round? eg:
- start and end date should be within the round dates
- scope should be a subset of round scopes
- vaccines should be a subset of round vaccines

quang-le committed:
Polio-1556: add translations

quang-le committed:
Polio-1556: add and update utils

- make baseUrl optional in useTabs
- Add useTableState hook
- export polio viruses
- fix query key in useSaveCampaigns

quang-le committed:
POLIO-1556: make ScopeInput reusable

- Pass campaign as prop i.o getting it via formik context
- Add useFilteredDistricts hook to allow filtering on different scopes
- Pass availableVaccines to scopes map to enable more flexible vaccine selection

quang-le committed:
POLIO-1556: add subActivities tab
- Add new components
- update polio dialog tabs

quang-le committed:
POLIO-1556: add PagintedResponse TS type

quang-le committed:
Merge branch 'main' into POLIO-1556_sub_activities_front_end

quang-le committed:
Merge branch 'main' into POLIO-1540-create-new-tab-sub-activities-for-all-campaigns-polio-and-non-polio

quang-le committed:
POLIO-1556: Add loading spinner on subactivity save

- prevent modal from closing when clicking Confirm
- Show spinner while mutation is performed
- close modal onSuccess. This allows the mutation hook to invalidate the query key and make sure the data is refetched

quang-le committed:
chore: remove dead code in LinkToInstance

quang-le committed:
POLIO-1556: code review

- fix spacing between input fields

quang-le committed:
Merge pull request #1340 from BLSQ/POLIO-1556_sub_activities_front_end

Polio-1556 sub activities front end
